packages/SwingSet/src/kernel/vat-warehouse.js

```diff
@@ -278,29 +447,50 @@ export function makeVatWarehouse({
   /** @type { string | undefined } */
   let lastVatID;
 
-  /** @type {(vatID: string, kd: KernelDeliveryObject, d: VatDeliveryObject, vs: VatSlog) => Promise<import('@agoric/swingset-liveslots').VatDeliveryResult> } */
+  /** @type {(vatID: string, kd: KernelDeliveryObject, d: VatDeliveryObject, vs: VatSlog) => Promise<VatDeliveryResult> } */
   async function deliverToVat(vatID, kd, vd, vs) {
     await applyAvailabilityPolicy(vatID);
     lastVatID = vatID;
 
     const recreate = true; // PANIC in the failure case
     // create the worker and replay the transcript, if necessary
-    const { manager } = await ensureVatOnline(vatID, recreate);
+    const {
+      manager,
+      syscallHandler: origHandler,
+      options,
+    } = await ensureVatOnline(vatID, recreate);
+    assert(origHandler);
 
     // then log the delivery so it appears after transcript replay
     const vatKeeper = kernelKeeper.provideVatKeeper(vatID);
     const crankNum = kernelKeeper.getCrankNumber();
     const deliveryNum = vatKeeper.nextDeliveryNum(); // increments
     /** @type { SlogFinishDelivery } */
-    const slogFinish = vs.delivery(crankNum, deliveryNum, kd, vd);
+    const finishSlog = vs.delivery(crankNum, deliveryNum, kd, vd);
+
+    // wrap the syscallHandler with a syscall recorder
+    const { syscallHandler, getSyscalls } = recordSyscalls(origHandler);
+    assert(syscallHandler);
 
     // make the delivery
-    const deliveryResult = await manager.deliver(vd);
+    const deliveryResult = await manager.deliver(vd, syscallHandler);
     insistVatDeliveryResult(deliveryResult);
+    finishSlog(deliveryResult); // log the delivery results
+
+    // TODO: if the dispatch failed for whatever reason, and we choose to
+    // destroy the vat, change what we do with the transcript here.
+    if (options.useTranscript) {
+      // record transcript entry
+      /** @type { import('../types-external.js').TranscriptDeliveryResults} */
+      const tdr = { status: deliveryResult[0] };
+      const transcriptEntry = { d: vd, sc: getSyscalls(), r: tdr };
+      vatKeeper.addToTranscript(transcriptEntry);
+    }
 
-    // log the delivery results, and return to caller for evaluation
-    slogFinish(deliveryResult);
-    return deliveryResult;
+    // TODO: if per-vat policy decides it wants a BOYD or heap snapshot,
```

Comment:
Btw, doing those without going through the runqueue likely makes more sense given our current model that these would be internal details of the vat's implementation.

Reply:
Yeah, one option I'm considering is to have the `deliverToVat` return results include a "please snapshot now" flag, and then `processDeliveryMessage` can see it and turn around and ask the warehouse to save a snapshot. The reason is that `deliverToVat` might be called by `processUpgradeVat`, which does *not* want to make a snapshot, because we're about to abandon the worker anyways.

Another reason is that `processDeliveryMessage` might see the results and decide to terminate the vat, in which case we also wouldn't want a snapshot. Both are exception cases, though, and maybe avoiding the useless snapshot is an optimization that isn't worth the code complexity it would require.

Maybe a better approach would be for `deliverToVat()` to accept an `inhibitSnapshot` option, provided by `processUpgradeVat` but not by `processDeliveryMessage`. That would avoid one of the cases, but not the terminated-vats-don't-need-snapshots case.

The most compelling argument for putting them in the run-queue (or more likely a higher-priority queue, like how we do GC actions and BOYD now) is if we thought we could batch the snapshot writes up into an end-of-block section. It depends a lot on our runPolicy and how expensive we mark snapshot writes, but we know for sure that it's a waste to snapshot a single vat more than once during a block. Just like the Nagle algorithm tries to batch together as many TCP writes as it can into a single IP frame, but doesn't know the future so it has to use a timer, our ideal behavior is to mark the vat as needing a snapshot during its delivery, but then not actually write the snapshot until later (but still soon). I don't really want the kernel to be so aware of block boundaries, but there ought to be a way to express this in the runPolicy, so the kernel can know that it will still have enough bean/computron budget to perform the snapshot writes later, and it knows when "later" is.

I guess in the long run, that means `runPolicy` wants to address a sequence of modes. Not just "do cranks" followed by "stop", but more like "do cranks", then "do snapshots", then stop. You get X beans to do cranks, then you get Y beans to do snapshots (and maybe you can do more cranks if you don't have any snapshots to do, but if those trigger snapshots then you push them onto the queue for next time), then you stop.

Huh, that sounds like I'm trying to put it *more* onto the run-queue. But not today.

Reply:
I'm trying to think of it with CDP in mind. The vat is in charge of snapshotting and recovering its state, and simply gets a queue of "external" deliveries. At some point (and deterministically), the vat decides its time to BOYD and snapshot.

I can see how a delivery that splats should preempt an immediate BOYD and snapshot, but that can fairly easily be worked out.

I don't fully understand the upgrade argument. I thought we weren't going to do any stopVat anymore? And if we do, it can be a special delivery that just forces a BOYD without snapshot, but no other work. However one could argue that creating a snapshot when doing an upgrade is actually useful if something goes wrong with the restart: we can simply respin the vat from the snapshot!

Regarding the runPolicy scheduling, yes that would possibly prevent batching at the end of the block. But I suppose that block informed scheduling is intrinsically incompatible with CDP. For now maybe we could have a similar vatRunPolicy provided by the host which informs whether the vat should perform a BOYD+snapshot after a delivery. Or as you describe a way for the run policy to decide what action to do next (snapshot or crank). I think in both cases this argues for BOYD+snapshot to not be part of the normal runqueue.

Regarding the multiple snapshot per block, in all cases we can simply avoid that by defining the snapshot compute interval to be higher than the block compute interval.